sdk/keyvault/TROUBLESHOOTING.md

```diff
@@ -0,0 +1,141 @@
+
+#### Operation Not Permitted
+
+You may see an error similar to:
+
+```text
+Azure.RequestFailedException : Service request failed.
+Status: 403 (Forbidden)
+
+Content:
+{"error":{"code":"Forbidden","message":"Operation decrypt is not permitted on this key.","innererror":{"code":"KeyOperationForbidden"}}}
+```
+
+The operation and inner `code` may vary, but the rest of the text will indicate which operation is not permitted.
+This error indicates that the authenticated application nor users does not have permissions to perform that operation,
+though the cause may vary.
+
+1. Check that the application or user has the appropriate permissions:
+   * [Access policies](https://docs.microsoft.com/azure/key-vault/general/assign-access-policy) (Key Vault)
+   * [Role-Based Access Control (RBAC)](https://docs.microsoft.com/azure/key-vault/general/rbac-guide) (Key Vault and Managed HSM)
+2. If the appropriate permissions are assigned to your application or user, make sure you are authenticating as that user.
+   If using the [DefaultAzureCredential] a different credential might've been used than one you expected.
+   [Enable logging](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/identity/Azure.Identity/README.md#logging)
+   and you will see which credential the [DefaultAzureCredential] used as shown below, and why previously-attempted credentials
+   were rejected.
+
+   ```text
+   [Informational] DefaultAzureCredential credential selected: Azure.Identity.EnvironmentCredential
+   ```
+
+#### Access Denied to First Party Service
+
+You may see an error similar to:
+
+```text
+Azure.RequestFailedException : Service request failed.
+Status: 403 (Forbidden)
+
+Content:
+{"error":{"code":"Forbidden","message":"Access denied to first party service. ...","innererror":{"code":"AccessDenied"}}}
+```
+
+The error `message` may also contain the tenant ID (`tid`) and application ID (`appid`). This error may occur because:
+
+1. You have the **Allow trust services** option enabled and are trying to access the Key Vault from a service not on
+   [this list](https://docs.microsoft.com/azure/key-vault/general/overview-vnet-service-endpoints#trusted-services) of
+   trust services.
+2. You are authenticated against a Microsoft Account (MSA) in Visual Studio or another credential provider. See
+   [above](#operation-not-permitted) for troubleshooting steps.
+
+## Other Service Errors
+
+To troubleshoot additional HTTP service errors not described below,
+visit [Azure Key Vault REST API Error Codes](https://docs.microsoft.com/azure/key-vault/general/rest-error-codes).
+
+### HTTP 429: Too Many Requests
+
+If you get an exception or see logs that describe HTTP 429, you may be making too many requests to Key Vault too quickly.
+
+Possible solutions include:
+
+1. Use a singleton for any `CertificateClient`, `KeyClient`, or `SecretClient` in your application for a single Key Vault.
```

Comment:
Perhaps link to any general docs we have here about singleton clients being recommended in general?

Reply:
Not aware of any, and there are plenty of different configuration libraries out there people use. If you have something in mind I welcome input.

Reply:
I suppose I could link to one we maintain as an example, making clear it is only that: an example.

Reply:
We have it as the recommended pattern in the blog article [Lifetime management for Azure SDK .NET clients](https://devblogs.microsoft.com/azure-sdk/lifetime-management-and-thread-safety-guarantees-of-azure-sdk-net-clients/#client-lifetime).